src/core/mapping/base_mapper.cc

```diff
@@ -207,6 +207,28 @@ void BaseMapper::select_task_options(const MapperContext ctx,
                                      const LegionTask& task,
                                      TaskOptions& output)
 {
+  LegateProjectionFunctor* key_functor = nullptr;
+  size_t idx                           = 0;
+  for (auto& req : task.regions) {
+    // read-only broadcasted instances should be collective instances without tags
+    bool ro_broadcasted = (req.projection == 0 && req.privilege == LEGION_READ_ONLY);
+    bool is_reduction   = (req.privilege == LEGION_REDUCE) && (req.projection != 0);
+    bool is_collective  = (is_reduction || ro_broadcasted);
+
+    // std::cout << "IRINA DEBUG inside of the select_task_options) task = " << task.get_task_name()
+    //           << " col = ? " << is_collective << ", reduction = ?" << is_reduction << std::endl;
+    if (is_collective) {
+      output.check_collective_regions.insert(idx);
+    } else if (req.tag == LEGATE_CORE_KEY_STORE_TAG) {
+      key_functor = find_legate_projection_functor(req.projection);
+      if (key_functor != nullptr && key_functor->is_collective()) {
+        output.check_collective_regions.insert(idx);
+        // std::cout <<"IRINA DEBUG is collective in mapper "<< task.get_task_name()<<std::endl;
+      }
+    }
+    idx++;
+  }
+
```

Comment:
```suggestion
  for (uint32_t idx = 0; idx < task.regions.size(); ++idx) {
    auto& req = task.regions[idx];
    if (find_legate_projection_functor(req.projection)->is_collective())
      output.check_collective_regions.insert(idx);
  }
```

Reply:
@magnatelee :did you remove 214-216, 220-222  intentionally? Without it reduction and broadcasted instances are not marked as collective.

Reply:
I forgot a condition in the if statement. Here's the right one:
```
for (uint32_t idx = 0; idx < task.regions.size(); ++idx) {
    auto& req = task.regions[idx];
    if (req.handle_type == LEGION_SINGULAR_PROJECTION || find_legate_projection_functor(req.projection)->is_collective())
      output.check_collective_regions.insert(idx);
  }
```
Note that your original code is incorrect because of two reasons: 1) it mis-categorizes a read-only requirement with an identity projection as a broadcasting requirement, 2) it misses a reduction requirement on a region, because that requirement has 0 for its `projection`.